src/analysis/convex/combination.lean

```diff
@@ -207,7 +217,38 @@ begin
       finset.centroid_weights_apply, zero_lt_one] }
 end
 
--- TODO : Do we need other versions of the next lemma?
+lemma convex_hull_range_eq_exists_affine_combination (v : ι → E) :
+  convex_hull R (range v) = { x | ∃ (s : finset ι) (w : ι → R)
+    (hw₀ : ∀ i ∈ s, 0 ≤ w i) (hw₁ : s.sum w = 1), s.affine_combination v w = x } :=
+begin
+  refine subset.antisymm (convex_hull_min _ _) _,
+  { intros x hx,
+    obtain ⟨i, hi⟩ := set.mem_range.mp hx,
+    refine ⟨{i}, function.const ι (1 : R), by simp, by simp, by simp [hi]⟩, },
+  { rw convex,
+    rintros x y ⟨s, w, hw₀, hw₁, rfl⟩ ⟨s', w', hw₀', hw₁', rfl⟩ a b ha hb hab,
+    let W : ι → R := λ i, (if i ∈ s then a * w i else 0) + (if i ∈ s' then b * w' i else 0),
+    have hW₁ : (s ∪ s').sum W = 1,
+    { rw [sum_add_distrib, ← sum_subset (subset_union_left s s'),
+        ← sum_subset (subset_union_right s s'), sum_ite_of_true _ _ (λ i hi, hi),
+        sum_ite_of_true _ _ (λ i hi, hi), ← mul_sum, ← mul_sum, hw₁, hw₁', ← add_mul, hab, mul_one];
+      intros i hi hi';
+      simp [hi'], },
+    refine ⟨s ∪ s', W, _, hW₁, _⟩,
+    { rintros i -,
+      by_cases hi : i ∈ s;
+      by_cases hi' : i ∈ s';
+      simp [hi, hi', add_nonneg, mul_nonneg ha (hw₀ i _), mul_nonneg hb (hw₀' i _)], },
+    { simp_rw [affine_combination_eq_linear_combination (s ∪ s') v _ hW₁,
+        affine_combination_eq_linear_combination s v w hw₁,
+        affine_combination_eq_linear_combination s' v w' hw₁', add_smul, sum_add_distrib],
+      rw [← sum_subset (subset_union_left s s'), ← sum_subset (subset_union_right s s')],
+      { simp only [ite_smul, sum_ite_of_true _ _ (λ i hi, hi), mul_smul, ← smul_sum], },
+      { intros i hi hi', simp [hi'], },
+      { intros i hi hi', simp [hi'], }, }, },
+  { rintros x ⟨s, w, hw₀, hw₁, rfl⟩,
+    exact affine_combination_mem_convex_hull hw₀ hw₁, },
+end
```

Comment:
I felt very sad that all those lemmas are duplicated so I tried to deduced this version from the existing lemma. But of course I got a proof that is longer... I'll paste this proof here anyway. If someone has courage maybe more `finset.sum` or affine geometry lemmas could be extracted to make it shorter. But I'm ready to merge your version if nobody has courage (however something is really wrong, all those proofs shouldn't be so painful).
```lean
lemma finset.sum_image'' {α β γ : Type*} [add_comm_monoid γ] (s : finset α) (f : α → β) (g : α → γ) :
 ∑ a in s, g a = ∑ b in s.image f, (∑ a in s, if f a = b then g a else 0) :=
begin
  rw finset.sum_image' g,  
  intros b b_in,  
  rw sum_filter
end

example (v : ι → E) :
  convex_hull R (range v) = { x | ∃ (s : finset ι) (w : ι → R)
    (hw₀ : ∀ i ∈ s, 0 ≤ w i) (hw₁ : s.sum w = 1), s.affine_combination v w = x } :=
begin
  rw convex_hull_eq,
  ext x,
  dsimp,
  split,
  { rintro ⟨J, s, w, p, hw, ⟨hw', mem, H⟩⟩,
    obtain ⟨j, j_in⟩ : s.nonempty,
    { by_contra H',
      rw [finset.not_nonempty_iff_eq_empty] at H',
      rw [H', sum_empty] at hw',
      exact zero_ne_one hw' },
    cases mem j j_in with i hi,
    haveI : nonempty ι := ⟨i⟩,
    dsimp [set.range] at mem,
    choose! f hf using mem,
    use finset.image f s,
    have h_sum : ∑ (i : ι) in image f s, ∑ (j : J) in s, ite (f j = i) (w j) 0 = 1,
    by rwa ← s.sum_image'',
    refine ⟨λ i, ∑ j in s, if f j = i then w j else 0, _, _, _⟩,
    { intros i i_in,
      apply sum_nonneg,
      intros j j_in,
      split_ifs ; simp [hw j j_in] },
    { exact h_sum },
    { rw affine_combination_eq_center_mass,
      { rw finset.center_mass_eq_of_sum_1 _ _ h_sum,
        rw finset.center_mass_eq_of_sum_1 _ _ hw' at H,
        rw ← H,
        dsimp,
        simp_rw [sum_smul, ite_smul, zero_smul],
        have : ∀ j i, ite (f j = i) (w j • v i) 0 = ite (f j = i) (w j • v (f j)) 0,
        { intros j i, 
          split_ifs with h ; simp [h] },
        simp_rw this,
        rw ← s.sum_image'' f,
        apply sum_congr rfl,
        intros j j_in,
        rw [hf j j_in] },
      { exact h_sum } } },
   { rintro ⟨s, w, w_nonneg, w_sum, rfl⟩,
    use [ι, s, w, v, w_nonneg, w_sum],
    simp [affine_combination_eq_center_mass w_sum] }
end
```

Reply:
I agree; also thanks for working out this new proof!

However I am inclined to stick with what I have. This is mostly because I think the right way to tidy up the mess here is to remove the (in my opinion) bad `center_mass` definition and instead to restate everything in terms of `finset.affine_combination`.

With that in mind I was motivated to come up with a proof that would be optimised for ease of such a future refactor, which is more or less what I've done here. In addition I started out reusing the existing lemmas but then decided (apparently correctly, in view of your proof) that it would be harder to write, harder to maintain, and longer.

Incidentally, my proof could be improved by building out a proper API for working with the fact that an affine combination of affine combinations is an affine combination but I felt I'd never get there if I took this detour.

I should also say that there is also a WIP PR #9268 which should be a step in the right direction.

Reply:
I agree, I pasted my code there to archive it somewhere just in case.